### PR TITLE
fix text-generation gpt2 bug

### PR DESCRIPTION
**Pull request by ZhaiFeiyue**
# What does this PR do?

as title

with this PR, text-generation with GPT2 could generate meaningful outputs

image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by ZhaiFeiyue** [comment]
@regisss with this PR, gpt2 with FP32 works well, but if enable BF16, the outputs' quality seems not good not like our tests before, I will look into BF16 issue later after enable GPT-J and OPT

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -434,3 +517,126 @@ def custom_forward(*inputs):
...
+        head_mask: Optional[torch.FloatTensor] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        encoder_attention_mask: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        token_idx=None,
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt2/modeling_gpt2.py:591]:
```suggestion
        token_idx: Optional[torch.Tensor] = None,
```

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 2]
```diff
@@ -434,3 +517,126 @@ def custom_forward(*inputs):
         attentions=all_self_attentions,
         cross_attentions=all_cross_attentions,
     )
+
+
+class GaudiGPT2LMHeadModel(GPT2LMHeadModel):
+    def __init__(self, config):
+        super().__init__(config)
+        self.lm_head_chunks = []
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt2/modeling_gpt2.py:525]: Is this used anywhere? Or is it for a future PR?

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 3]
```diff
@@ -434,3 +517,126 @@ def custom_forward(*inputs):
         attentions=all_self_attentions,
         cross_attentions=all_cross_attentions,
     )
+
+
+class GaudiGPT2LMHeadModel(GPT2LMHeadModel):
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt2/modeling_gpt2.py:522]:
```suggestion
class GaudiGPT2LMHeadModel(GPT2LMHeadModel):
    """
    Copied from GPT2LMHeadModel: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py
    The only differences are:
    - add new args token_idx
    """
```

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 4]
```diff
@@ -187,6 +189,7 @@ def forward(
         encoder_attention_mask: Optional[torch.FloatTensor] = None,
         use_cache: Optional[bool] = False,
         output_attentions: Optional[bool] = False,
+        token_idx=None,
```
[the remark this answers is not in the document]
Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py]: done

File: examples/text-generation/checkpoint_utils.py [hunk 5]
```diff
@@ -64,3 +64,15 @@ def model_is_bloom(config):
     Checks if the given config belongs to a BLOOM-like model.
     """
     return "bloom" in config.architectures[0].lower()
+
+
+def model_is_optimized(config):
+    """
+    Checks if the given config belongs to a model in optimum/habana/transformers/models, which has a
+    new input token_idx
```
[the remark this answers is not in the document]
Comment (ZhaiFeiyue) [reply examples/text-generation/checkpoint_utils.py]: done

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 6]
```diff
@@ -246,11 +326,13 @@ def gaudi_gpt2_forward(
     output_attentions: Optional[bool] = None,
     output_hidden_states: Optional[bool] = None,
     return_dict: Optional[bool] = None,
+    token_idx=None,
```
[the remark this answers is not in the document]
Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py]: done

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 7]
```diff
@@ -231,6 +241,76 @@ def forward(
...
+    self,
+    hidden_states: Optional[Tuple[torch.FloatTensor]],
+    layer_past: Optional[Tuple[torch.Tensor]] = None,
+    attention_mask: Optional[torch.FloatTensor] = None,
+    head_mask: Optional[torch.FloatTensor] = None,
+    encoder_hidden_states: Optional[torch.Tensor] = None,
+    encoder_attention_mask: Optional[torch.FloatTensor] = None,
+    use_cache: Optional[bool] = False,
+    output_attentions: Optional[bool] = False,
+    token_idx=None,
```
[the remark this answers is not in the document]
Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py]: done

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py (hunk 1 above)
Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py]: done

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py (hunk 3 above)
Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py:522]: done

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py (hunk 2 above)
Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py]: not used code, has been removed

**Comment by ZhaiFeiyue** [comment]
@regisss code has been updated base on your comments

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 8]
```diff
@@ -434,3 +517,131 @@ def custom_forward(*inputs):
...
+
+class GaudiGPT2LMHeadModel(GPT2LMHeadModel):
+    """
+    Copied from GPT2LMHeadModel: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py
+    The only differences are:
+    - add new args token_idx
+    """
+
+    def __init__(self, config):
+        super().__init__(config)
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt2/modeling_gpt2.py:530]: We can remove this block then

**Review by regisss — COMMENTED** [review]
Thanks for updating this PR @ZhaiFeiyue! Let's also remove the `__init__` in `GaudiGPT2LMHeadModel` since `self.lm_head_chunks` was removed. And then it will be good to merge!

**Review by regisss — APPROVED** [review]
LGTM!

**Pull request merged**